lib/tlog/play.c

```diff
@@ -80,172 +309,60 @@ tlog_play_create_json_reader(struct tlog_errs **perrs,
-        grc = tlog_es_json_reader_create(&reader, baseurl, query, 10);
+        /* Create es reader */
+        grc = tlog_play_create_es_json_reader(perrs, &reader, reader_conf);
         if (grc != TLOG_RC_OK) {
-            tlog_errs_pushc(perrs, grc);
-            tlog_errs_pushs(perrs, "Failed creating the Elasticsearch reader");
             goto cleanup;
         }
 #ifdef TLOG_JOURNAL_ENABLED
     } else if (strcmp(str, "journal") == 0) {
-        struct json_object *conf_journal;
-        int64_t since = 0;
-        int64_t until = INT64_MAX;
-        size_t i;
-
         /* Get journal reader conf container */
-        if (json_object_object_get_ex(conf, "journal", &conf_journal)) {
-            /* Get the "since" timestamp */
-            if (json_object_object_get_ex(conf_journal, "since", &obj)) {
-                since = json_object_get_int64(obj);
-                if (since < 0) {
-                    since = 0;
-                } else if ((uint64_t)since > UINT64_MAX/1000000) {
-                    since = UINT64_MAX/1000000;
-                }
-            } else {
-                since = 0;
-            }
-
-            /* Get the "until" timestamp */
-            if (json_object_object_get_ex(conf_journal, "until", &obj)) {
-                until = json_object_get_int64(obj);
-                if (until < 0) {
-                    until = 0;
-                } else if ((uint64_t)until > UINT64_MAX/1000000) {
-                    until = UINT64_MAX/1000000;
-                }
-            } else {
-                until = INT64_MAX;
-            }
-
-            /* Get the match array, if any */
-            if (json_object_object_get_ex(conf_journal, "match", &obj)) {
-                str_list = calloc(json_object_array_length(obj) + 1,
-                                  sizeof(*str_list));
-                if (str_list == NULL) {
-                    grc = TLOG_GRC_ERRNO;
-                    tlog_errs_pushc(perrs, grc);
-                    tlog_errs_pushs(perrs,
-                                    "Failed allocating systemd journal match list");
-                    goto cleanup;
-                }
-                for (i = 0; (int)i < (int)json_object_array_length(obj); i++) {
-                    str_list[i] = json_object_get_string(
-                                    json_object_array_get_idx(obj, i));
-                    if (!tlog_journal_match_sym_is_valid(str_list[i])) {
-                        grc = TLOG_RC_FAILURE;
-                        tlog_errs_pushf(
-                            perrs,
-                            "Systemd journal match symbol #%zu \"%s\" "
-                            "is invalid",
-                            i + 1, str_list[i]);
-                        goto cleanup;
-                    }
-                }
-            }
+        if (!json_object_object_get_ex(conf, str, &reader_conf)) {
+            tlog_errs_pushs(perrs, "Journal reader parameters are not specified");
```

Comment:
This breaks running `tlog-play` without journal reader parameters (we need integration tests ASAP to catch these). I.e. as just `tlog-play -r journal`. This normally is not useful, but in general should be supported. You can fix this by allowing passing `NULL` `conf` to the reader-creating function. All the `json_*` functions would then act as if the node you're retrieving doesn't exist, and it should work OK. Also, while you're updating and moving around these lines, please make sure that they are within 78 chars long. Thanks!